test/handshake-memfail.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2016-2025 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * We need access to the deprecated low level HMAC APIs for legacy purposes
+ * when the deprecated calls are not hidden
+ */
+#ifndef OPENSSL_NO_DEPRECATED_3_0
+# define OPENSSL_SUPPRESS_DEPRECATED
+#endif
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openssl/crypto.h>
+#include <openssl/opensslconf.h>
+#include <openssl/bio.h>
+#include <openssl/crypto.h>
+#include <openssl/ssl.h>
+#include <openssl/ocsp.h>
+#include <openssl/srp.h>
+#include <openssl/txt_db.h>
+#include <openssl/aes.h>
+#include <openssl/rand.h>
+#include <openssl/core_names.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/provider.h>
+#include <openssl/param_build.h>
+#include <openssl/x509v3.h>
+#include <openssl/dh.h>
+#include <openssl/engine.h>
+
+#include "helpers/ssltestlib.h"
+#include "testutil.h"
+#include "testutil/output.h"
+#include "internal/nelem.h"
+#include "internal/tlsgroups.h"
+#include "internal/ktls.h"
+#include "internal/ssl_unwrap.h"
+#include "../ssl/ssl_local.h"
+#include "../ssl/record/methods/recmethod_local.h"
+#include "filterprov.h"
+
+/**
+ * @brief Global static variables for certificate and key handling.
+ *
+ * These variables store the paths and context used for managing
+ * certificates and private keys in the application.
+ *
+ * - opmode: "count" or "run" to define operational mode
+ * - certsdir: Directory containing trusted certificates.
+ * - cert:     Path to the certificate file in use.
+ * - privkey:  Path to the private key file.
+ * - libctx:   OpenSSL library context used for cryptographic operations.
+ * - mcount:   Number of mallocs counted
+ * - rcount:   Number of reallocs counted
+ * - fcount:   Number of frees counted
+ * - scount:   Number of mallocs counted prior to workload
+ */
+static char *opmode = NULL;
+static char *certsdir = NULL;
+static char *cert = NULL;
+static char *privkey = NULL;
+static OSSL_LIB_CTX *libctx = NULL;
+static int mcount, rcount, fcount, scount;
+
```

Comment:
nit: can `opmode` and certsdir`  be made local variables in `main()`?

also `libctx` can be made local to test functions. `do_handshake()` then can receive libctx as argument.